src/gui/styles_dialog.c

```diff
@@ -910,11 +934,19 @@ GtkWidget *dt_gui_style_content_dialog(char *name, const int imgid)
     gtk_widget_set_halign(da, GTK_ALIGN_CENTER);
     gtk_widget_set_app_paintable(da, TRUE);
     gtk_box_pack_start(GTK_BOX(ht), da, TRUE, TRUE, 0);
-    _preview_data_t *data = g_malloc(sizeof(_preview_data_t));
-    g_strlcpy(data->style_name, name, sizeof(data->style_name));
-    data->imgid = imgid;
-    g_signal_connect_data(G_OBJECT(da), "draw", G_CALLBACK(_preview_draw), data, (GClosureNotify)g_free, 0);
+    g_strlcpy(data.style_name, name, sizeof(data.style_name));
+    data.imgid = imgid;
+    data.first_draw = TRUE;
+    g_free(data.hash);
```

Comment:
I have introduce a way to free history hashes (#13934). This will probably be better here too. You just need to copy the hash here and use a single dt_history_hash_free() below.